src/main/scala/cromwell/binding/Scope.scala

```diff
@@ -1,14 +1,151 @@
+        case s: Ast if s.getName == AstNodeName.Scatter => Option(Scatter(s, scopeIndexes.getOrElse(classOf[Scatter], 0), parent))
+
+        /* Cases need to be added here to support new scopes */
+        case _ => None
+      }
+      //Generate and set children recursively
+      scope foreach { sc =>
+        //Increment index if needed
+        scopeIndexes.put(sc.getClass, scopeIndexes.getOrElse(sc.getClass, 0) + 1)
+        sc.setChildren(getChildrenList(node) flatMap { generateScopeTree(_, namespaces, tasks, wdlSyntaxErrorFormatter, Option(sc)) })
+      }
+      scope
+    }
+
+    generateScopeTree(node, namespaces, tasks, wdlSyntaxErrorFormatter, None)
+      .getOrElse(throw new UnsupportedOperationException("Input node cannot be instantiated as a Scope"))
+  }
+
+  /**
+   * Generate a workflow from an Ast. The ast must be one of a workflow.
+   * @throws UnsupportedOperationException if the ast is not a workflow ast
+   * @return a workflow
+   */
+  def generateWorkflow(ast: Ast, namespaces: Seq[WdlNamespace], tasks: Seq[Task], wdlSyntaxErrorFormatter: WdlSyntaxErrorFormatter): Workflow = {
+    ast.getName match {
+      case AstNodeName.Workflow => generateTree(ast, namespaces, tasks, wdlSyntaxErrorFormatter).asInstanceOf[Workflow]
+      case nonWorkflowAst => throw new UnsupportedOperationException(s"Ast is not a 'Workflow Ast' but a '$nonWorkflowAst Ast'")
+    }
+  }
+
+  /**
+   * Collect Calls from a Seq of Scopes.
+   * @param scopes scopes to loop through
+   * @return Scopes instances that are Calls
+   */
+  def collectCalls(scopes: Seq[Scope]): Seq[Call] = scopes collect { case s: Call => s }
+
+  /**
+   * Collect all Calls from the given scope.
+   * @param scopes scope to gather Calls from
+   * @param calls for recursivity. Should be passed Nil in most cases.
+   * @return all Calls inside the scope
+   */
+  @tailrec
+  def collectAllCalls(scopes: Seq[Scope], calls: Seq[Call]): Seq[Call] = scopes match {
+    case Nil => calls
+    case l => collectAllCalls(l.flatMap(_.children), calls ++ collectCalls(l))
+  }
+
+  /**
+   * Collect Scatters from a Seq of Scopes.
+   * @param scopes scopes to loop through
+   * @return Scopes instances that are Scatters
+   */
+  def collectScatters(scopes: Seq[Scope]): Seq[Scatter] = scopes collect { case s: Scatter => s }
+
+  /**
+   * Collect all Scatters from the given scope.
+   * @param scopes scope to gather Scatters from
+   * @param scatters for recursivity. Should be passed Nil in most cases.
+   * @return all Scatters inside the scope
+   */
+  @tailrec
+  def collectAllScatters(scopes: Seq[Scope], scatters: Seq[Scatter]): Seq[Scatter] = scopes match {
+    case Nil => scatters
+    case l => collectAllScatters(l.flatMap(_.children), scatters ++ collectScatters(l))
+  }
+
+  @tailrec
+  def fullyQualifiedNameBuilder(scope: Option[Scope], fqn: String, fullDisplay: Boolean): String = scope match {
+    case None => fqn.tail //Strip away the first "." of the name
```

Comment:
`fqn.tail` will throw if fqn is the empty string, which doesn't look impossible?